segmentstore/server/src/test/java/io/pravega/segmentstore/server/DataRecoveryTestUtils.java

```
@@ -0,0 +1,188 @@
+                    IteratorArgs.builder().fetchTimeout(TIMEOUT).build()).join();
+
+            // Add all segments present in the container metadata in a set.
+            Set<TableKey> segmentsInMD = new HashSet<>();
+            it.forEachRemaining(k -> segmentsInMD.addAll(k.getEntries()), EXECUTOR_SERVICE).join();
+
+            for (SegmentProperties segment : segments) {
+                long len = segment.getLength();
+                boolean isSealed = segment.isSealed();
+                String segmentName = segment.getName();
+
+                /*
+                    1. segment exists in both metadata and storage, re-create it
+                    2. segment only in metadata, delete
+                    3. segment only in storage, re-create it
+                 */
+                segmentsInMD.remove(TableKey.unversioned(getTableKey(segmentName)));
+                container.getStreamSegmentInfo(segment.getName(), TIMEOUT)
+                        .thenAccept(e -> {
+                            container.createStreamSegment(segmentName, len, isSealed)
+                                    .exceptionally(ex -> {
+                                        log.error("Exception occurred while creating segment", ex);
+                                        return null;
+                                    }).join();
+                        })
+                        .exceptionally(e -> {
+                            log.error("Got an exception on getStreamSegmentInfo", e);
+                            if (Exceptions.unwrap(e) instanceof StreamSegmentNotExistsException) {
+                                container.createStreamSegment(segmentName, len, isSealed)
+                                        .exceptionally(ex -> {
+                                            log.error("Exception occurred while creating segment", ex);
+                                            return null;
```

You should fail the whole process here instead of just logging a message.

---

OK.